File: backend/routes/air_conditioner.py
"""
空调模块 API 路由
负责人：lzp
功能：空调控制、温湿度监控、历史数据查询
"""

from flask import Blueprint, jsonify, request
from database import (
    get_recent_data, get_devices, get_latest_data,
    upsert_ac_state, get_ac_state, get_all_acs, 
    insert_ac_event, get_ac_events
)

# 创建蓝图
air_conditioner_bp = Blueprint('air_conditioner', __name__)


@air_conditioner_bp.route("/devices", methods=["GET"])
def devices():
    """获取所有设备列表"""
    print(f"收到设备列表请求 - Method: {request.method}, Origin: {request.headers.get('Origin')}")
    data = get_devices()
    return jsonify(data)


@air_conditioner_bp.route("/history")
def history():
    """获取所有设备的历史数据"""
    limit = request.args.get('limit', 100, type=int)
    data = get_recent_data(limit=limit)
    return jsonify(data)


@air_conditioner_bp.route("/history/<device_id>")
def history_by_device(device_id):
    """获取指定设备的历史数据"""
    limit = request.args.get('limit', 100, type=int)
    data = get_recent_data(device_id=device_id, limit=limit)
    return jsonify(data)


@air_conditioner_bp.route("/latest/<device_id>")
def latest(device_id):
    """获取指定设备的最新数据"""
    data = get_latest_data(device_id)
    if data:
        return jsonify(data)
    else:
        return jsonify({"error": "设备未找到"}), 404


# ==================== 空调控制 API ====================

@air_conditioner_bp.route("/ac", methods=["GET"])
def list_acs():
    """获取所有空调列表"""
    acs = get_all_acs()
    return jsonify(acs)


@air_conditioner_bp.route("/ac/<ac_id>", methods=["GET"])
def ac_state(ac_id):
    """获取空调状态"""
    state = get_ac_state(ac_id)
    if state:
        # 获取最新的温湿度数据
        device_id = state.get('device_id', 'room1')
        latest_data = get_latest_data(device_id)
        if latest_data:
            state['current_temp'] = latest_data['temperature']
            state['current_humidity'] = latest_data['humidity']
            # 更新数据库中的当前温湿度
            upsert_ac_state(
                ac_id, 
                current_temp=latest_data['temperature'],
                current_humidity=latest_data['humidity']
            )
        return jsonify(state)
    return jsonify({"error": "空调未找到"}), 404


@air_conditioner_bp.route("/ac/<ac_id>/control", methods=["POST"])
def control_ac(ac_id):
    """控制空调（开关、温度设置等）"""
    body = request.get_json(force=True) or {}
    
    # 获取当前状态
    current_state = get_ac_state(ac_id)
    
    # 提取控制参数
    power = body.get('power')  # True/False
    target_temp = body.get('target_temp')  # 目标温度
    mode = body.get('mode')  # cool/heat/fan
    fan_speed = body.get('fan_speed')  # low/medium/high/auto
    device_id = body.get('device_id', 'room1')
    
    # 记录旧值（用于事件日志）
    old_power = current_state['power'] if current_state else False
    old_temp = current_state['target_temp'] if current_state else None
    
    # 更新空调状态
    upsert_ac_state(
        ac_id=ac_id,
        device_id=device_id,
        power=power,
        mode=mode,
        target_temp=target_temp,
        fan_speed=fan_speed
    )
    
    # 记录事件
    if power is not None and power != old_power:
        event_type = 'power_on' if power else 'power_off'
        insert_ac_event(
            ac_id=ac_id,
            event_type=event_type,
            old_value=str(old_power),
            new_value=str(power),
            detail=f"空调{'开启' if power else '关闭'}"
        )
    
    if target_temp is not None and target_temp != old_temp:
        insert_ac_event(
            ac_id=ac_id,
            event_type='temp_change',
            old_value=str(old_temp) if old_temp else None,
            new_value=str(target_temp),
            detail=f"目标温度设置为 {target_temp}°C"
        )
    
    return jsonify({
        "status": "success",
        "message": "空调控制成功",
        "ac_id": ac_id
    })


@air_conditioner_bp.route("/ac/<ac_id>/events", methods=["GET"])
def ac_events(ac_id):
    """获取空调事件历史"""
    limit = request.args.get('limit', 50, type=int)
    events = get_ac_events(ac_id, limit)
    return jsonify(events)

File: frontend/air-conditioner.js
// API 基础地址
const API_BASE = 'http://localhost:5000';

// 初始化图表
const chart = echarts.init(document.getElementById('chart'));
let currentDevice = 'all';

// 房间固定排序
const ROOM_ORDER = ['living_room', 'bedroom1', 'bedroom2', 'kitchen', 'study'];

// 排序函数
function sortByRoomOrder(items, idField = 'device_id') {
    return items.sort((a, b) => {
        const indexA = ROOM_ORDER.indexOf(a[idField]);
        const indexB = ROOM_ORDER.indexOf(b[idField]);
        // 如果房间不在列表中，放到最后
        if (indexA === -1 && indexB === -1) return 0;
        if (indexA === -1) return 1;
        if (indexB === -1) return -1;
        return indexA - indexB;
    });
}

// 图表配置
const option = {
    title: { 
        text: '温湿度实时监控',
        left: 'center'
    },
    tooltip: {
        trigger: 'axis',
        axisPointer: {
            type: 'cross'
        }
    },
    legend: {
        data: ['温度', '湿度'],
        top: 50
    },
    grid: {
        left: '3%',
        right: '4%',
        bottom: '3%',
        containLabel: true
    },
    xAxis: {
        type: 'category',
        boundaryGap: false,
        data: []
    },
    yAxis: [
        {
            type: 'value',
            name: '温度 (°C)',
            position: 'left',
            axisLabel: {
                formatter: '{value} °C'
            }
        },
        {
            type: 'value',
            name: '湿度 (%)',
            position: 'right',
            axisLabel: {
                formatter: '{value} %'
            }
        }
    ],
    series: [
        {
            name: '温度',
            type: 'line',
            smooth: true,
            yAxisIndex: 0,
            data: [],
            itemStyle: {
                color: '#FF6B6B'
            }
        },
        {
            name: '湿度',
            type: 'line',
            smooth: true,
            yAxisIndex: 1,
            data: [],
            itemStyle: {
                color: '#4ECDC4'
            }
        }
    ]
};

chart.setOption(option);

// 加载设备列表
async function loadDevices() {
    try {
        console.log('正在请求设备列表:', `${API_BASE}/devices`);
        console.log('当前页面 Origin:', window.location.origin);

        // 添加详细的请求配置和调试信息
        const requestOptions = {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
            mode: 'cors', // 明确指定 CORS 模式
            credentials: 'omit' // 不发送凭据
        };

        console.log('请求配置:', requestOptions);
        // 添加时间戳避免缓存
        const url = `${API_BASE}/devices?_t=${Date.now()}`;
        const response = await fetch(url, requestOptions);
        console.log('响应状态:', response.status, response.statusText);
        console.log('响应头:', [...response.headers.entries()]);

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const devices = await response.json();
        console.log('获取到设备数据:', devices);

        // 按固定顺序排序
        const sortedDevices = sortByRoomOrder(devices, 'device_id');

        const select = document.getElementById('deviceSelect');
        select.innerHTML = '';  // 清空选项

        sortedDevices.forEach(device => {
            const option = document.createElement('option');
            option.value = device.device_id;
            // 优先显示中文房间名，如果没有则显示device_id
            const displayName = device.room_name || device.device_id;
            option.textContent = `${displayName} (${device.data_count} 条数据)`;
            select.appendChild(option);
        });

        // 默认选择第一个设备
        if (sortedDevices.length > 0) {
            currentDevice = sortedDevices[0].device_id;
        }

        updateStatus('设备列表加载成功', 'success');
    } catch (error) {
        console.error('加载设备列表失败:', error);
        console.error('错误详情:', error.message);
        updateStatus(`加载设备失败: ${error.message}`, 'error');
    }
}

// 加载历史数据
async function loadHistory(deviceId) {
    try {
        updateStatus('加载中...', 'loading');
        
        // 如果没有指定设备，使用当前设备
        if (!deviceId) {
            deviceId = currentDevice || 'room1';
        }
        
        const baseUrl = `${API_BASE}/history/${deviceId}?limit=50`;
        
        // 添加时间戳避免缓存
        const url = baseUrl + `&_t=${Date.now()}`;
        
        console.log('正在请求历史数据:', url);
        
        const requestOptions = {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
            mode: 'cors',
            credentials: 'omit'
        };
        
        const response = await fetch(url, requestOptions);
        console.log('历史数据响应状态:', response.status, response.statusText);
        console.log('历史数据响应头:', [...response.headers.entries()]);
        
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        
        const data = await response.json();
        console.log('获取到历史数据:', data.length, '条记录');
        
        if (data.length === 0) {
            updateStatus('暂无数据', 'warning');
            return;
        }
        
        // 按时间排序（升序）
        data.sort((a, b) => new Date(a.timestamp) - new Date(b.timestamp));
        
        // 提取数据
        const times = data.map(d => new Date(d.timestamp).toLocaleTimeString());
        const temperatures = data.map(d => d.temperature);
        const humidities = data.map(d => d.humidity);
        
        // 更新图表
        option.xAxis.data = times;
        option.series[0].data = temperatures;
        option.series[1].data = humidities;
        chart.setOption(option);
        
        // 更新统计信息
        const latest = data[data.length - 1];
        document.getElementById('currentTemp').textContent = `${latest.temperature}°C`;
        document.getElementById('currentHum').textContent = `${latest.humidity}%`;
        document.getElementById('dataCount').textContent = data.length;
        
        // 计算舒适指数
        updateComfortIndex(latest.temperature, latest.humidity);
        
        updateStatus('数据加载成功', 'success');
        
    } catch (error) {
        console.error('加载数据失败:', error);
        console.error('错误详情:', error.message);
        updateStatus(`加载数据失败: ${error.message}`, 'error');
    }
}

// 更新舒适指数
function updateComfortIndex(temp, hum) {
    let comfortLevel = '舒适';
    let comfortColor = '#28a745';
    
    if (temp > 28 || temp < 16) {
        comfortLevel = '不舒适';
        comfortColor = '#dc3545';
    } else if (hum > 80 || hum < 30) {
        comfortLevel = '一般';
        comfortColor = '#ffc107';
    }
    
    document.getElementById('comfortIndex').textContent = comfortLevel;
    document.getElementById('comfortIndex').style.color = comfortColor;
}

// 更新状态显示
function updateStatus(message, type = 'info') {
    const statusEl = document.getElementById('status');
    statusEl.textContent = message;
    statusEl.className = `status status-${type}`;
}

// 空调控制
// 根据当前设备动态生成空调ID
function getACId() {
    // 如果是 'all' 或未设置，默认使用 room1
    let device = currentDevice;
    if (!device || device === 'all') {
        device = 'room1';
    }
    return `ac_${device}`;
}

// 当前选中的模式
let currentMode = 'cool';

function setupACControls() {
    const acButtons = document.querySelectorAll('.ac-btn');
    const modeButtons = document.querySelectorAll('.mode-btn');
    
    // 加载初始空调状态
    loadACStatus();
    
    // 模式切换按钮事件
    modeButtons.forEach(button => {
        button.addEventListener('click', async () => {
            const mode = button.getAttribute('data-mode');
            currentMode = mode;
            
            // 更新按钮样式
            modeButtons.forEach(btn => btn.classList.remove('active'));
            button.classList.add('active');
            
            // 如果空调正在运行，立即应用新模式
            const acId = getACId();
            const state = await getACState();
            if (state && state.power) {
                await controlAC(true, state.target_temp, mode);
            } else {
                updateStatus(`已选择${getModeDisplayName(mode)}模式`, 'success');
            }
        });
    });
    
    // 温度控制按钮事件
    acButtons.forEach(button => {
        button.addEventListener('click', async () => {
            if (button.id === 'acOff') {
                // 关闭空调
                await controlAC(false, null, null);
            } else {
                const temp = button.getAttribute('data-temp');
                // 开启空调并设置温度（使用当前选择的模式）
                await controlAC(true, parseFloat(temp), currentMode);
            }
        });
    });
    
    // 定时刷新空调状态
    setInterval(loadACStatus, 5000);
}

// 获取空调状态（不显示UI）
async function getACState() {
    try {
        const acId = getACId();
        const response = await fetch(`${API_BASE}/ac/${acId}?_t=${Date.now()}`, {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            cache: 'no-cache'
        });
        
        if (response.ok) {
            return await response.json();
        }
        return null;
    } catch (error) {
        console.error('获取空调状态失败:', error);
        return null;
    }
}

// 获取模式显示名称
function getModeDisplayName(mode) {
    const modeNames = {
        'cool': '❄️ 制冷',
        'heat': '🔥 制热',
        'fan': '💨 送风',
        'dehumidify': '💧 除湿'
    };
    return modeNames[mode] || mode;
}

// 加载空调状态
async function loadACStatus() {
    try {
        const acId = getACId();
        const response = await fetch(`${API_BASE}/ac/${acId}?_t=${Date.now()}`, {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            cache: 'no-cache'
        });
        
        if (response.ok) {
            const state = await response.json();
            const acStatus = document.getElementById('acStatus');
            const acModeEl = document.getElementById('acMode');
            
            if (state.power) {
                acStatus.textContent = `运行中 (目标: ${state.target_temp}°C, 当前: ${state.current_temp}°C)`;
                acStatus.style.color = '#28a745';
            } else {
                acStatus.textContent = '已关闭';
                acStatus.style.color = '#dc3545';
            }
            
            // 显示当前模式
            if (state.mode) {
                acModeEl.textContent = getModeDisplayName(state.mode);
                currentMode = state.mode;
                
                // 更新模式按钮高亮
                document.querySelectorAll('.mode-btn').forEach(btn => {
                    btn.classList.remove('active');
                    if (btn.getAttribute('data-mode') === state.mode) {
                        btn.classList.add('active');
                    }
                });
            }
        }
    } catch (error) {
        console.error('加载空调状态失败:', error);
    }
}

// 控制空调
async function controlAC(power, targetTemp, mode = null) {
    try {
        updateStatus('正在控制空调...', 'loading');
        
        // 获取正确的设备ID
        let device = currentDevice;
        if (!device || device === 'all') {
            device = 'room1';
        }
        
        const body = {
            power: power,
            device_id: device,
            mode: mode || currentMode  // 使用传入的模式或当前选择的模式
        };
        
        if (targetTemp !== null) {
            body.target_temp = targetTemp;
        }
        
        const acId = getACId();
        const response = await fetch(`${API_BASE}/ac/${acId}/control`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
            body: JSON.stringify(body)
        });
        
        if (response.ok) {
            const result = await response.json();
            
            const acStatus = document.getElementById('acStatus');
            const acModeEl = document.getElementById('acMode');
            
            if (power) {
                const modeDisplay = getModeDisplayName(mode || currentMode);
                acStatus.textContent = `运行中 (${targetTemp}°C)`;
                acStatus.style.color = '#28a745';
                acModeEl.textContent = modeDisplay;
                updateStatus(`空调已设置为 ${modeDisplay} ${targetTemp}°C`, 'success');
            } else {
                acStatus.textContent = '已关闭';
                acStatus.style.color = '#dc3545';
                updateStatus('空调已关闭', 'success');
            }
            
            // 刷新状态
            setTimeout(loadACStatus, 1000);
        } else {
            // 尝试获取错误详情
            const errorText = await response.text();
            console.error('服务器响应:', response.status, errorText);
            throw new Error(`控制失败 (${response.status}): ${errorText.substring(0, 100)}`);
        }
    } catch (error) {
        console.error('控制空调失败:', error);
        updateStatus(`控制失败: ${error.message}`, 'error');
    }
}

// 事件监听
document.getElementById('deviceSelect').addEventListener('change', (e) => {
    currentDevice = e.target.value;
    loadHistory(currentDevice);
    // 切换设备时也重新加载空调状态
    loadACStatus();
});

document.getElementById('refreshBtn').addEventListener('click', () => {
    loadHistory(currentDevice);
});

// 自动刷新（每10秒）
setInterval(() => {
    loadHistory(currentDevice);
}, 10000);

// 响应式图表
window.addEventListener('resize', () => {
    chart.resize();
});

// 初始化
loadDevices();
loadHistory();
setupACControls();